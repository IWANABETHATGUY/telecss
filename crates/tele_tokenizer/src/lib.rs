#![warn(missing_docs)]

//! CSS Tokenizer

use std::{iter::Peekable, str::CharIndices};

mod inner_state;
mod token;

use inner_state::*;
use tele_utils::*;
use token::{Pos, Token, TokenType};

#[derive(Debug)]
pub struct Tokenizer<'s> {
  source: &'s str,
  iter: Peekable<CharIndices<'s>>,
  line: usize,
  colnmu: usize,
  offset: usize,
  pre_cursor: Pos,
  cursor: Pos,
  current_state: State,
  buffer: Vec<char>,
  tokens: Vec<Token>,
}

impl<'s> From<&'s str> for Tokenizer<'s> {
  fn from(source: &'s str) -> Self {
    Tokenizer {
      source,
      iter: source.char_indices().peekable(),
      // position related
      line: 1,
      colnmu: 1,
      offset: 0,
      pre_cursor: Pos::default(),
      cursor: Pos::default(),
      // state machine related
      current_state: State::Initial,
      buffer: Vec::with_capacity(100),
      tokens: Vec::new(),
    }
  }
}

impl<'s> Tokenizer<'s> {
  pub fn tokenize(&mut self) -> &mut Vec<Token> {
    while let Some(&(offset, c)) = self.iter.peek() {
      self.current_state = match self.current_state {
        State::Initial => match c {
          c if is_whitespace(c) => State::WhiteSpace,
          '.' => State::MayBeNumber,
          _ => {
            self.advance(offset);
            State::Initial
          }
        },
        State::WhiteSpace => {
          if is_whitespace(c) {
            if is_newline(c) {
              self.line += 1;
              self.colnmu = 1;
            }
            self.consume(offset, c);
            State::WhiteSpace
          } else {
            self.emit(TokenType::WhiteSpace);
            State::Initial
          }
        }
        State::MayBeNumber => {
          self.consume(offset, c);
          if is_digit(c) {
            State::EOF
          } else {
            self.emit(TokenType::Delim);
            State::Initial
          }
        }
        State::EOF => break,
      };
    }

    &mut self.tokens
  }

  fn consume(&mut self, offset: usize, c: char) {
    // consume
    self.advance(offset);
    self.buffer.push(c);
  }

  fn advance(&mut self, offset: usize) {
    self.iter.next();
    self.colnmu += 1;
    self.offset = offset;
  }

  fn get_cursor(&self) -> Pos {
    Pos {
      offset: self.offset,
      line: self.line,
      column: self.colnmu,
    }
  }

  fn emit(&mut self, token_type: TokenType) {
    let cur_cursor = self.get_cursor();

    let mut token = Token::default();
    token.token_type = token_type;
    token.start_pos = self.pre_cursor;
    token.end_pos = cur_cursor;
    token.content.append(&mut self.buffer);

    self.pre_cursor = cur_cursor;
    self.tokens.push(token)
  }
}
