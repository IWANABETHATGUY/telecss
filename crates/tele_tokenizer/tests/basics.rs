use insta::assert_json_snapshot;
use tele_tokenizer::*;

#[test]
fn it_works() {
  let mut tokenizer: Tokenizer = ".foo { color: red; }".into();
  assert_json_snapshot!(tokenizer.tokenize());
}

#[test]
fn comments() {
  let mut tokenizer: Tokenizer = " /** I'm Comment * */ ".into();
  assert_json_snapshot!(tokenizer.tokenize());
}

#[test]
fn comma() {
  let mut tokenizer: Tokenizer = ".foo, .bar {}".into();
  assert_json_snapshot!(tokenizer.tokenize());
}

#[test]
fn unexpected_eof_comments() {
  let mut tokenizer: Tokenizer = " /** I'm Comment *".into();
  let res = tokenizer.tokenize();
  assert_eq!(res, Err(Error::from(ErrorKind::UnexpectedEOF)));

  let mut tokenizer: Tokenizer = " /** I'm Comment".into();
  let res = tokenizer.tokenize();
  assert_eq!(res, Err(Error::from(ErrorKind::UnexpectedEOF)));
}
