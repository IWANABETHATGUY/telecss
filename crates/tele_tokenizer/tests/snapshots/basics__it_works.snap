---
source: crates/tele_tokenizer/tests/basics.rs
assertion_line: 7
expression: tokenizer.tokenize()

---
{
  "Ok": [
    {
      "token_type": "Delim",
      "start_pos": {
        "offset": 0,
        "line": 1,
        "column": 1
      },
      "end_pos": {
        "offset": 0,
        "line": 1,
        "column": 2
      },
      "content": [
        "."
      ]
    },
    {
      "token_type": "Ident",
      "start_pos": {
        "offset": 0,
        "line": 1,
        "column": 2
      },
      "end_pos": {
        "offset": 3,
        "line": 1,
        "column": 5
      },
      "content": [
        "f",
        "o",
        "o"
      ]
    },
    {
      "token_type": "WhiteSpace",
      "start_pos": {
        "offset": 3,
        "line": 1,
        "column": 5
      },
      "end_pos": {
        "offset": 4,
        "line": 1,
        "column": 6
      },
      "content": [
        " "
      ]
    },
    {
      "token_type": "LeftCurlyBracket",
      "start_pos": {
        "offset": 4,
        "line": 1,
        "column": 6
      },
      "end_pos": {
        "offset": 5,
        "line": 1,
        "column": 7
      },
      "content": [
        "{"
      ]
    },
    {
      "token_type": "WhiteSpace",
      "start_pos": {
        "offset": 5,
        "line": 1,
        "column": 7
      },
      "end_pos": {
        "offset": 6,
        "line": 1,
        "column": 8
      },
      "content": [
        " "
      ]
    },
    {
      "token_type": "Ident",
      "start_pos": {
        "offset": 6,
        "line": 1,
        "column": 8
      },
      "end_pos": {
        "offset": 11,
        "line": 1,
        "column": 13
      },
      "content": [
        "c",
        "o",
        "l",
        "o",
        "r"
      ]
    },
    {
      "token_type": "Colon",
      "start_pos": {
        "offset": 11,
        "line": 1,
        "column": 13
      },
      "end_pos": {
        "offset": 12,
        "line": 1,
        "column": 14
      },
      "content": [
        ":"
      ]
    },
    {
      "token_type": "WhiteSpace",
      "start_pos": {
        "offset": 12,
        "line": 1,
        "column": 14
      },
      "end_pos": {
        "offset": 13,
        "line": 1,
        "column": 15
      },
      "content": [
        " "
      ]
    },
    {
      "token_type": "Ident",
      "start_pos": {
        "offset": 13,
        "line": 1,
        "column": 15
      },
      "end_pos": {
        "offset": 16,
        "line": 1,
        "column": 18
      },
      "content": [
        "r",
        "e",
        "d"
      ]
    },
    {
      "token_type": "SemiColon",
      "start_pos": {
        "offset": 16,
        "line": 1,
        "column": 18
      },
      "end_pos": {
        "offset": 17,
        "line": 1,
        "column": 19
      },
      "content": [
        ";"
      ]
    },
    {
      "token_type": "WhiteSpace",
      "start_pos": {
        "offset": 17,
        "line": 1,
        "column": 19
      },
      "end_pos": {
        "offset": 18,
        "line": 1,
        "column": 20
      },
      "content": [
        " "
      ]
    },
    {
      "token_type": "RightCurlyBracket",
      "start_pos": {
        "offset": 18,
        "line": 1,
        "column": 20
      },
      "end_pos": {
        "offset": 19,
        "line": 1,
        "column": 21
      },
      "content": [
        "}"
      ]
    },
    {
      "token_type": "EOF",
      "start_pos": {
        "offset": 19,
        "line": 1,
        "column": 21
      },
      "end_pos": {
        "offset": 19,
        "line": 1,
        "column": 21
      },
      "content": []
    }
  ]
}
